### ISSUE TITLE
The accuracy of using Graph to implement Bidirectional RNN for classification is note correct

### ISSUE THREAD

**Issue by icemansina** [fchollet/keras#1369]
Currently, I am using Bidirectional RNN/LSTM(BRNN/BLSTM) for sequence classification (return_sequence= True, i.e., classification of each time-step). According to the official provided example, the BRNN is required to implemented by Graph (while the BidirectionalRNN API does not support mask input). Besides, before mini-batch input, I do a preprocessing (pad with 0) to make the input to (batch_size, input_length, input_features), but it seems that no explicit mask input for keras. So I listed my implementation as 

`modellstm = Graph()`
`modellstm.add_input(name='input', input_shape=(MAX_SEQ_LENGTH, input_dim), dtype='float')`
`modellstm.add_node(Masking(mask_value=0), name='masked_input', input='input')`
`modellstm.add_node(LSTM(10, return_sequences=True), name='forward', input='masked_input')`
`modellstm.add_node(LSTM(10, go_backwards= True, return_sequences=True), name='backward', input='masked_input')`
`modellstm.add_node(Dropout(0.5), name='dropout', inputs=['forward', 'backward'])`
`modellstm.add_node(TimeDistributedDense(N_CLASSES, activation='softmax'), name='dense', input='dropout')`
`modellstm.add_output(name='output', input='dense')`
`modellstm.compile('Adadelta', {'output': 'categorical_crossentropy'})`

As no **show_accuracy=True** for modellstm.fit (fit of Graph does not support), the accuracy 
`acc = accuracy((y_train_accuracy),  np.argmax(np.array(modellstm.predict({'input': X_train}, batch_size=BATCH_SIZE)['output']), axis=-1))` doesn't take the input_mask into consideration. Besides, for example, **without explicit input_mask**, **the output result of classification np.argmax([0 0 0 0])=0 (i.e., class=1) for input np.array([0 0 ... 0]), which is indeed wrong**. Actually input_mask for input np.array([0 0 ... 0]) should be **0**, which means all 0 input and output should not be calculated for accuracy. How could I modify me code to accomplish this code. Thanks a lot for your nice assistance. @fchollet @farizrahman4u

**Comment by meitcher** [comment]
This is not only for Bidirectional, in fact, all RNNs with Masking had this behavior in version 0.2.0 (now I don't know). In my project, I created a callback for calculate the accuracy using an unpadding method. Try with:

```python
callbacks = [ModelCheckpoint(filepath=weights_file, verbose=1, save_best_only=True)]
callbacks.append(EpochAccuracy((samples, target, batch_size), eval_func=calculate_accuracy))

self.model.fit({'input': samples, 'output': target}, nb_epoch=nb_epoch, batch_size=batch_size, callbacks=callbacks)

class EpochAccuracy(keras.callbacks.Callback):
    def __init__(self, data, eval_func):
        self.samples, self.target, self.batch_size = data
        self.eval = eval_func

    def on_epoch_end(self, epoch, logs={}):
        candidates = np.argmax(self.model.predict({'input':self.samples}, verbose=0, batch_size=self.batch_size)['output'], axis=-1)
        candidates = unpad_sequences(candidates, mask_value=0)
        accuracy = self.eval(candidates, self.target)
        logger.info("Accuracy after epoch {}: {}".format(epoch, accuracy))

```
The ``unpad_sequences`` looks like:
```python
def unpad_sequences(padded_sequences, mask_value=0):
    unpadded = []
    for sequence in padded_sequences:
        # XXX: might remove mask_values from the middle of the sequence
        unpadded.append(sequence[sequence != mask_value].tolist())
    return unpadded
```
Another way to unpad the sequences is using the original sequence as mask.

**Comment by icemansina** [comment]
@meitcher Thanks for your reply.
In keras v0.3.0. the accuracy is calculated as:
`if class_mode == "categorical":`
 `      train_accuracy = K.mean(K.equal(K.argmax(self.y, axis=-1), K.argmax(self.y_train, axis=-1)))`
 `       test_accuracy = K.mean(K.equal(K.argmax(self.y, axis=-1), K.argmax(self.y_test, axis=-1)))`
and the mask is still not considered. 

In your implementation, I think it is also required to do the unpad_sequences operations to target, i.e.,

 `candidates = unpad_sequences(candidates, mask_value=0)`
`self.target = unpad_sequences(self.target, mask_value=0)`
`accuracy = self.eval(candidates, self.target)`

or

`self.eval(y_preds[input_mask_flat], y_true[input_mask_flat])`

How do you think about it?

**Comment by meitcher** [comment]
In #957 is showed that is also possible to use the sample_weights in fit() method, which is a simple way to do mask in the output.

I think that it is not necessary, since that doesn't make sense to pad the target values... But it depends on your implementation.

**Comment by danieldk** [comment]
I ran into the same problem and found out that the accuracy function does not take masking into account by reading the source code... Is this considered to be a bug in the accuracy function or the documentation? I'm asking, because I could try to whip up a pull request ;).
